homeassistant/helpers/script.py

```diff
@@ -47,107 +49,126 @@ def __init__(self, hass: HomeAssistant, sequence, name: str=None,
         self.last_action = None
         self.can_cancel = any(CONF_DELAY in action for action
                               in self.sequence)
-        self._lock = threading.Lock()
-        self._unsub_delay_listener = None
+        self._async_unsub_delay_listener = None
         self._template_cache = {}
 
     @property
     def is_running(self) -> bool:
         """Return true if script is on."""
         return self._cur != -1
 
-    def run(self, variables: Optional[Sequence]=None) -> None:
+    def run(self, variables=None):
         """Run script."""
-        with self._lock:
-            if self._cur == -1:
-                self._log('Running script')
-                self._cur = 0
-
-            # Unregister callback if we were in a delay but turn on is called
-            # again. In that case we just continue execution.
-            self._remove_listener()
-
-            for cur, action in islice(enumerate(self.sequence), self._cur,
-                                      None):
-
-                if CONF_DELAY in action:
-                    # Call ourselves in the future to continue work
-                    def script_delay(now):
-                        """Called after delay is done."""
-                        self._unsub_delay_listener = None
-                        self.run(variables)
-
-                    delay = action[CONF_DELAY]
-
-                    if isinstance(delay, template.Template):
-                        delay = vol.All(
-                            cv.time_period,
-                            cv.positive_timedelta)(
-                                delay.render())
-
-                    self._unsub_delay_listener = track_point_in_utc_time(
+        run_coroutine_threadsafe(
+            self.async_run(variables), self.hass.loop).result()
+
+    @asyncio.coroutine
+    def async_run(self, variables: Optional[Sequence]=None) -> None:
+        """Run script.
+
+        Returns a coroutine.
```

Comment:
Interesting comment. I was under the impression that no matter what you return from a coroutine, it will always return a generator object when called.

However, I do realize that I remembered it wrong from the Python docs that I was trying to mimic: it should be: `This method is a coroutine.`